src/aggregator/client/writer_mgr_test.go

```diff
@@ -65,6 +73,189 @@ func TestWriterManagerAddInstancesSingleRef(t *testing.T) {
+// It first overrides the connect and write functions of the TcpClient and
+// and makes the writeFn block on the context.Context. An instance created
+// with this overridden connection options struct creates a slow writer.
+// After the writer/instance is created, we re-override the connect/write
+// functions in the connection options struct and replace them with no-op
+// functions. These writers are called normal writers. Instances added after
+// updating the opts with the re-overridden functions creates normal writers.
+// The test creates several normal writers and one slow writer.
+// It then begins a loop of writing one payload to every writer: slow and normal.
+// The first Flush() invoked on the slow writer blocks the thread and so
+// We initiate every write-loop in a separate goroutine. This mimics the newly
+// updated ReportSnapshot() semantics where every second it is invoked in a
+// separate goroutine.
+// We keep track of attempted writes and completed writes of the normal
+// clients by way of counters: writesInitialized and writesCompleted.
+// These counters are counted and checked in a separate single goroutine to
+// to avoid any data races.
+// As soon as all write-loops are done meaning that all normal writers have
+// finished writing all that had to write, the context is cancelled. This unblocks
+// the slow writer and then we perform the necessary validations.
+// NB(shankar.nair): An issue with this test is that the slow writer times out
+// and only then the test is able to complete. This is because the goroutine
+// worker pool in the write manager has an issue of head-of-line blocking where
+// a normal writer could be blocked behind the slow writer. Therefore, since this
+// test waits for the normal writer to be done, it first has to wait for the
+// slow writer to finish via timeout. This makes this test a bit flaky and
+// can fail in such situations.
+func TestWriterManagerMultipleWriters(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	wrInitCh := make(chan bool)
+	wrCompCh := make(chan bool)
+	doneCh := make(chan bool)
+
+	writesCompleted := 0
+	writesInitiated := 0
+
+	ctx := context.Background()
+	ctx, cancelFn := context.WithCancel(ctx)
+	defer cancelFn()
+
+	slowMockConn := NewMockConn(ctrl)
+	slowMockConn.EXPECT().Write(gomock.Any()).DoAndReturn(func(b []byte) (n int, err error) {
+		// Block till all normal writers have finished writing
+		// all write loops below
+		select {
+		case <-time.After(1 * time.Second):
+		case <-ctx.Done():
+		}
+		return len(b), nil
+	}).AnyTimes()
+	slowMockConn.EXPECT().SetWriteDeadline(gomock.Any()).AnyTimes()
+
+	slowWriterDialerFn := func(c context.Context, network string, address string) (net.Conn, error) {
+		return slowMockConn, nil
+	}
+
+	normalMockConn := NewMockConn(ctrl)
+	normalMockConn.EXPECT().Write(gomock.Any()).DoAndReturn(func(b []byte) (n int, err error) {
+		// signal write completion and return immediately
+		wrCompCh <- true
```

Comment:
I think you want a sync.WaitGroup here :) https://pkg.go.dev/sync

Much simpler way of waiting for a group to finish.